In [1]:
import pandas as pd
import numpy as np
import seaborn as sns

In [2]:
data = pd.read_csv('../02_NeuralCTR/train.csv')

In [3]:
data.head()

,label,integer_feature_1,integer_feature_2,integer_feature_3,integer_feature_4,integer_feature_5,integer_feature_6,integer_feature_7,integer_feature_8,integer_feature_9,...,categorical_feature_17,categorical_feature_18,categorical_feature_19,categorical_feature_20,categorical_feature_21,categorical_feature_22,categorical_feature_23,categorical_feature_24,categorical_feature_25,categorical_feature_26
0,0,49.0,520.0,0.0,353.0,5.0,0.0,0.0,0,2,...,1f7fc70b,628f1b8d,9512c20b,c5362157,a9ad4e8e,2b6a4b2f,bc71cd9d,9f4102b0,ff654802,2f33557c
1,0,NaN,408.0,NaN,NaN,NaN,0.0,0.0,-1,1,...,NaN,a1eb1511,0a1d513f,7fdb0729,747f6ecc,62682372,NaN,401a16fe,30436bfc,b757e957
2,0,28.0,37.0,2.0,5.0,4.0,5.0,1.0,974,14,...,753da5f3,d9f758ff,9512c20b,682d90ba,09afda28,307a0f03,f942364c,16e6b497,57e36578,a7ab145d
3,0,1.0,603.0,1.0,NaN,1.0,0.0,0.0,13,1,...,NaN,b8170bba,108a0699,47849e55,73b3f46d,d994ba60,NaN,328e967b,ff654802,2ba8d787
4,0,NaN,525.0,3.0,8.0,126.0,1.0,0.0,10,11,...,753da5f3,b8170bba,55191c09,0e29b5da,0683bc6f,47b99d80,d2b125b2,00a688ec,30436bfc,b757e957


In [4]:
print(data['label'].dtype)

int64


In [5]:
type(list(data.columns))

list

In [6]:
categorical_feat = [i for i in data.columns if data[i].dtype == 'str']
numerical_feat = [i for i in data.columns if data[i].dtype == float or data[i].dtype == int]
numerical_feat = numerical_feat[1:]

In [7]:
## preprocessing and handle the NAN values

## I put median value from NAN in integer_feature and max occurance for categorical_features

for i in numerical_feat:
    data[i] = data[i].fillna(data[i].median())


for i in categorical_feat:
    data[i] = data[i].fillna('missing')



In [8]:
data.info()

<class 'pandas.DataFrame'>
RangeIndex: 40000 entries, 0 to 39999
Data columns (total 40 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   label                   40000 non-null  int64  
 1   integer_feature_1       40000 non-null  float64
 2   integer_feature_2       40000 non-null  float64
 3   integer_feature_3       40000 non-null  float64
 4   integer_feature_4       40000 non-null  float64
 5   integer_feature_5       40000 non-null  float64
 6   integer_feature_6       40000 non-null  float64
 7   integer_feature_7       40000 non-null  float64
 8   integer_feature_8       40000 non-null  int64  
 9   integer_feature_9       40000 non-null  int64  
 10  integer_feature_10      40000 non-null  float64
 11  integer_feature_11      40000 non-null  float64
 12  integer_feature_12      40000 non-null  float64
 13  integer_feature_13      40000 non-null  float64
 14  categorical_feature_1   40000 non-null  str    
 

In [9]:
categorical_data_unique = {}

for i in categorical_feat:
    categorical_data_unique[i] = data[i].unique().shape

In [10]:
categorical_data_unique

{'categorical_feature_1': (13191,),
 'categorical_feature_2': (3757,),
 'categorical_feature_3': (4968,),
 'categorical_feature_4': (1851,),
 'categorical_feature_5': (4253,),
 'categorical_feature_6': (3,),
 'categorical_feature_7': (3111,),
 'categorical_feature_8': (741,),
 'categorical_feature_9': (22,),
 'categorical_feature_10': (11585,),
 'categorical_feature_11': (5569,),
 'categorical_feature_12': (7723,),
 'categorical_feature_13': (9,),
 'categorical_feature_14': (904,),
 'categorical_feature_15': (2025,),
 'categorical_feature_16': (43,),
 'categorical_feature_17': (4,),
 'categorical_feature_18': (341,),
 'categorical_feature_19': (14,),
 'categorical_feature_20': (13509,),
 'categorical_feature_21': (9003,),
 'categorical_feature_22': (12629,),
 'categorical_feature_23': (5095,),
 'categorical_feature_24': (3504,),
 'categorical_feature_25': (36,),
 'categorical_feature_26': (24,)}

In [11]:
## Train test split

X = data.drop(columns=['label'])
y = data['label']

In [12]:
from sklearn.model_selection import train_test_split

X_train,X_test,y_train,y_test = train_test_split(
    X,y,test_size=0.2,random_state = 42
)

In [13]:
from sklearn.preprocessing import OneHotEncoder,StandardScaler

scaler = StandardScaler()

In [14]:
# scaling numerical data

X_train_num = scaler.fit_transform(X_train[numerical_feat])
X_test_num = scaler.transform(X_test[numerical_feat])

X_test_num.shape

(8000, 13)

In [15]:
## prepare categorical data

X_train_cat = X_train[categorical_feat].astype('str')
X_test_cat = X_test[categorical_feat].astype('str')

In [16]:
X_train_cat.head()

,categorical_feature_1,categorical_feature_2,categorical_feature_3,categorical_feature_4,categorical_feature_5,categorical_feature_6,categorical_feature_7,categorical_feature_8,categorical_feature_9,categorical_feature_10,...,categorical_feature_17,categorical_feature_18,categorical_feature_19,categorical_feature_20,categorical_feature_21,categorical_feature_22,categorical_feature_23,categorical_feature_24,categorical_feature_25,categorical_feature_26
14307,5dc16de2,d372df4b,62a5fee5,14841cc2,664ff944,6fcd6dcb,7dfd04ed,038e402d,2e4e821f,baac6e18,...,missing,b8170bba,9512c20b,39fffff8,a3e62678,1a7d6a4a,missing,2f01f6d1,991321ea,b757e957
17812,missing,2027c0d0,64fa3e43,a8e2f1c9,1315f676,6fcd6dcb,2d72123e,20aa0da1,25dd8f9a,missing,...,missing,7232d217,9512c20b,missing,missing,missing,missing,59c61d47,321935cd,6c730e3e
11020,7db3a66e,ea9a22db,48c07a22,6faef306,e755064d,6fcd6dcb,fad5de89,1c904173,2e4e821f,f72b2fb9,...,1f7fc70b,d9f758ff,9512c20b,8e0391a0,2798c0f9,3e29a205,e54bad85,348246ae,30436bfc,2f33557c
15158,e7b94768,1d52b21b,73369e23,386c49ee,4ff7b61d,6fcd6dcb,3d65ae52,8f551379,2e4e821f,79d81167,...,1f7fc70b,b8170bba,139221a3,df359cec,b5ed6108,e68b6fba,b8df7c07,3581f19f,ff654802,2ccea557
24990,436a9128,11074985,200258ea,f7f54f97,664ff944,6fcd6dcb,11525a1a,09f8a09d,2e4e821f,23d4c410,...,753da5f3,b8170bba,139221a3,1b15b20d,c89adb62,bffbac21,4fb3556e,941f47a2,ff654802,2f33557c


In [17]:
## creating DLRM for numerical features

from tensorflow.keras.models import Sequential, Model
from tensorflow.keras.layers import Dense,Input,Embedding,StringLookup,Dot,Concatenate
import tensorflow as tf


In [19]:
numerical_input = Input(shape=(X_train_num.shape[1],),name="numerical_input")

In [20]:
# implement layers of dense
embedding_dim = 16 ## we will convert str features in 16 dimensions features


dense_model = Sequential([
    Dense(64, activation="relu"),
    Dense(32, activation="relu"),
    Dense(embedding_dim, activation="relu")
])

dense = dense_model(numerical_input)

In [21]:
## embedding of features for categorical data

categorical_inputs = []
categorical_embeddings = []

In [22]:

for feature in categorical_feat:
    print("Processing:", feature)

    input_layer = Input(
        shape=(1,),
        dtype=tf.string,
        name=feature
    )

    lookup = StringLookup(
        num_oov_indices=1,
        output_mode="int"
    )

    lookup.adapt(X_train_cat[feature].astype(str).tolist())

    encoded = lookup(input_layer)

    embedding = Embedding(
        input_dim=lookup.vocabulary_size(),
        output_dim=16
    )(encoded)

    embedding = tf.keras.layers.Flatten()(embedding)

    categorical_inputs.append(input_layer)
    categorical_embeddings.append(embedding)

print("DONE")

Processing: categorical_feature_1
Processing: categorical_feature_2
Processing: categorical_feature_3
Processing: categorical_feature_4
Processing: categorical_feature_5
Processing: categorical_feature_6
Processing: categorical_feature_7
Processing: categorical_feature_8
Processing: categorical_feature_9
Processing: categorical_feature_10
Processing: categorical_feature_11
Processing: categorical_feature_12
Processing: categorical_feature_13
Processing: categorical_feature_14
Processing: categorical_feature_15
Processing: categorical_feature_16
Processing: categorical_feature_17
Processing: categorical_feature_18
Processing: categorical_feature_19
Processing: categorical_feature_20
Processing: categorical_feature_21
Processing: categorical_feature_22
Processing: categorical_feature_23
Processing: categorical_feature_24
Processing: categorical_feature_25
Processing: categorical_feature_26
DONE


In [23]:
## Concate the numerical feature and categorical feature

feature_vectors = [
    dense
] + categorical_embeddings

In [24]:
len(feature_vectors)

27

In [25]:
instructions = []

for i in range(len(feature_vectors)):
    for j in range(i+1,len(feature_vectors)):
        instruction = Dot(axes=1)([feature_vectors[i],feature_vectors[j]])
        instructions.append(instruction)

In [26]:
instructions_op = Concatenate()([dense]+instructions)

In [27]:
instructions_op.shape

(None, 367)

In [ ]:


x = Dense(64, activation='relu')(instructions_op)
x = Dense(32, activation='relu')(x)
output = Dense(1, activation='sigmoid')(x)

model = Model(inputs=[numerical_input] + categorical_inputs, outputs=output)

model.compile(
    optimizer='adam',
    loss='binary_crossentropy',
    metrics=['accuracy',tf.keras.metrics.AUC(name='roc_auc'),tf.keras.metrics.AUC(name='pr_auc',curve='PR')] ## without using only accuracy i also used auc
)

In [29]:
# model.summary()

In [30]:
train_inputs = {
    "numerical_input": tf.convert_to_tensor(
        X_train_num,
        dtype=tf.float32
    )
}

for feature in categorical_feat:
    train_inputs[feature] = tf.convert_to_tensor(
        X_train_cat[feature].values,
        dtype=tf.string
    )

test_inputs = {
    "numerical_input": tf.convert_to_tensor(
        X_test_num,
        dtype=tf.float32
    )
}

for feature in categorical_feat:
    test_inputs[feature] = tf.convert_to_tensor(
        X_test_cat[feature].values,
        dtype=tf.string
    )

In [31]:
print(y_train.dtype)
print(train_inputs["numerical_input"].dtype)
print(train_inputs["categorical_feature_1"].dtype)

int64
<dtype: 'float32'>
<dtype: 'string'>


In [32]:
from tensorflow.keras.callbacks import EarlyStopping

## Set up Early Stopping

early_stopping_callback = EarlyStopping(monitor='val_loss',patience=10,restore_best_weights=True)

In [33]:
history = model.fit(
    train_inputs,
    y_train,
    epochs=20,
    batch_size=32,
    validation_data=(test_inputs,y_test),
    callbacks=[early_stopping_callback]
)

Epoch 1/20
1000/1000 ━━━━━━━━━━━━━━━━━━━━ 31s 19ms/step - accuracy: 0.9676 - loss: 0.1515 - pr_auc: 0.0462 - roc_auc: 0.6205 - val_accuracy: 0.9694 - val_loss: 0.1302 - val_pr_auc: 0.0667 - val_roc_auc: 0.6987
Epoch 2/20
1000/1000 ━━━━━━━━━━━━━━━━━━━━ 16s 16ms/step - accuracy: 0.9682 - loss: 0.1047 - pr_auc: 0.2961 - roc_auc: 0.8750 - val_accuracy: 0.9679 - val_loss: 0.1582 - val_pr_auc: 0.0567 - val_roc_auc: 0.6225
Epoch 3/20
1000/1000 ━━━━━━━━━━━━━━━━━━━━ 16s 16ms/step - accuracy: 0.9838 - loss: 0.0482 - pr_auc: 0.7939 - roc_auc: 0.9778 - val_accuracy: 0.9595 - val_loss: 0.2114 - val_pr_auc: 0.0511 - val_roc_auc: 0.5847
Epoch 4/20
1000/1000 ━━━━━━━━━━━━━━━━━━━━ 16s 16ms/step - accuracy: 0.9922 - loss: 0.0222 - pr_auc: 0.9435 - roc_auc: 0.9949 - val_accuracy: 0.9499 - val_loss: 0.2801 - val_pr_auc: 0.0436 - val_roc_auc: 0.5693
Epoch 5/20
1000/1000 ━━━━━━━━━━━━━━━━━━━━ 18s 18ms/step - accuracy: 0.9966 - loss: 0.0102 - pr_auc: 0.9843 - roc_auc: 0.9979 - val_accuracy: 0.9352 - val_loss: 

In [34]:
data_test = pd.read_csv('../02_NeuralCTR/test.csv')

In [35]:
# Numerical features
for feature in numerical_feat:
    data_test[feature] = data_test[feature].fillna(
        data[feature].median()
    )

# Categorical features
for feature in categorical_feat:
    data_test[feature] = data_test[feature].fillna("missing")

# Separate features
X_test_new = data_test.drop(columns=["label"])
y_test_new = data_test["label"]

# Scale numerical features using the ALREADY fitted scaler
X_test_new_num = scaler.transform(
    X_test_new[numerical_feat]
)

# Keep categorical features as strings
X_test_new_cat = X_test_new[categorical_feat].astype(str)

In [36]:
test_inputs = [X_test_new_num]

for feature in categorical_feat:
    test_inputs.append(
        tf.convert_to_tensor(
            X_test_new_cat[feature].to_numpy().reshape(-1, 1),
            dtype=tf.string
        )
    )

In [37]:
from sklearn.metrics import (
    roc_auc_score,
    accuracy_score,
    average_precision_score,
    log_loss,
    f1_score,
    precision_score,
    recall_score,
    classification_report,
    confusion_matrix
)

In [45]:
# True labels
y_true = np.asarray(y_test_new).astype(int).ravel()

# Model predictions
y_pred = model.predict(test_inputs).ravel()


313/313 ━━━━━━━━━━━━━━━━━━━━ 2s 6ms/step


In [46]:
## finding best threshold for best f1 score

import numpy as np
from sklearn.metrics import f1_score, precision_score, recall_score

thresholds = np.arange(0.01, 0.51, 0.01)

best_threshold = 0
best_f1 = 0

for threshold in thresholds:
    y_pred_class = (y_pred >= threshold).astype(int)

    f1 = f1_score(y_true, y_pred_class)

    if f1 > best_f1:
        best_f1 = f1
        best_threshold = threshold

print("Best threshold:", best_threshold)
print("Best F1:", best_f1)

Best threshold: 0.06999999999999999
Best F1: 0.1478494623655914


In [47]:
# Default threshold
threshold = best_threshold
y_pred_class = (y_pred >= threshold).astype(int)

# Metrics
roc_auc = roc_auc_score(y_true, y_pred)
accuracy = accuracy_score(y_true, y_pred_class)
pr_auc = average_precision_score(y_true, y_pred)
logloss = log_loss(y_true, y_pred)
f1 = f1_score(y_true, y_pred_class)
precision = precision_score(y_true, y_pred_class)
recall = recall_score(y_true, y_pred_class)

print("ROC-AUC  :", roc_auc)
print("Accuracy :", accuracy)
print("PR-AUC   :", pr_auc)
print("Log Loss :", logloss)
print("F1 Score :", f1)
print("Precision:", precision)
print("Recall   :", recall)

print("\nClassification Report:")
print(classification_report(y_true, y_pred_class))

print("\nConfusion Matrix:")
print(confusion_matrix(y_true, y_pred_class))

ROC-AUC  : 0.7115641142451221
Accuracy : 0.8732
PR-AUC   : 0.08073274141379018
Log Loss : 0.13787220418453217
F1 Score : 0.1478494623655914
Precision: 0.09540329575021683
Recall   : 0.3283582089552239

Classification Report:
              precision    recall  f1-score   support

           0       0.97      0.89      0.93      9665
           1       0.10      0.33      0.15       335

    accuracy                           0.87     10000
   macro avg       0.53      0.61      0.54     10000
weighted avg       0.95      0.87      0.91     10000


Confusion Matrix:
[[8622 1043]
 [ 225  110]]


#### other scores are low but Accuracy is high beacause of the imbalanced training dataset. there are approx 96% data contains output is 0. for output 1 model remember the input and output.  that's why it scores low.